### DISCUSSION TITLE
Firejail alone is probably good enough for security and privacy

### DISCUSSION THREAD

**Discussion by amano-kenji** [netblue30/firejail#5465] [General]
While AppArmor can provide extra security, as long as you install small softwares from linux distribution repositories, you don't get much benefit from protecting system applications with AppArmor. How likely is it that a window manager or a terminal emulator is going to hack your computer or send your data to remote servers? They don't run remote codes.

On the other hand, big untrusted softwares that may or may not run arbitrary remote codes are web browsers and PDF readers and other user applications that can be sandboxed by firejail. A PDF reader can execute javascript in PDF.

Let's admit that true security is an illusion. Living on earth is not safe. Earthquakes, floods, hurricanes, and other natural disasters can destroy your house anytime. Even with SELinux, locking down the entire system is impossible. Often, hardwares and softwares promising total lockdown are spywares like Intel ME and AMD PSP. Anyone who promises total security wants control. Microsoft tried to use secure boot to prevent computers from replacing windows with linux. Government promises security but constantly threatens violence if people don't pay up or don't adhere to its arbitrarily changing rules without letting people consent to rules. Governments have killed many more people than all criminal organizations combined. Yet, instead of taking responsibility for our own security, we think we need governments to keep us safe. True security doesn't exist on earth. If you give up freedom to big tech or big government for illusion of security, you lose security. Lesson // Don't give up freedom for security.

But, you can get reasonable levels of security and privacy without compromising your natural right to do things without initiating violence. I think firejail is a good tradeoff. It is relatively straightfoward to use for applications that have firejail profiles.

I think creating an AppArmor profile for any system service that runs arbitrary remote code or directly connects to the internet is still worth my time. If you are trying to protect every system application with AppArmor or SELinux, you will be stuck creating and modifying profiles until your death. Perfection is impossible as long as you live in the physical universe.

Most of the time, protection from arbitrary code execution and on networked software is enough.

**Comment by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [comment]
Dear [amano-kenji], that is exactly my saying and what I mostly try to explain. 
I agree mostly, and there are no words that describe how awesome it is that you put this out here !

**Reply by amano-kenji** [reply]
While I think people can write an operating system that sandboxes applications by default, it doesn't exist, yet.

I don't have time to pursue perfect security.

**Reply by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [reply]
Dear @amano-kenji , I think you are wrong i.e.
SubgraphOS was one of many attempts supported by ThoughtWorks here where I am living),clear Linux another,....,etc.
By default, I guess you know that already from firejail the Linux is confining a lot of processes(threads) and their childs, even without seccomp. Also, with the principle of least privilege we already have in our OS, a lot of containment, that is why all the audits always have that many false positives. Just like you said "security" is relative and it depends on what you consider "sandboxed" ;) Post Script: As you know it is mostly unnecessary to firejail, google's version of the kernel is a good example, if it would not have proper containment(Seccomp) people wouldn't use android, and every router I know(even those from cisco) would be vulnerable ...

**Reply by amano-kenji** [reply]
Jonathan Blow already said linux doesn't put applications in sandbox by default.

Android sandboxes all appstore apps by default. I forgot about android and other mobile OSes.

So, I meant to say we don't have a desktop operating system where sandboxing is the default for applications.

**Reply by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [reply]
Well it depends on what you mean by "sandboxing", and I am pretty sure that some flavors of linux have that. Linux is a too broad term, but a random exe of a process is by default contained in every release version of the most popular flavors. If with linux you mean Arch Linux you are probably right, but that are not the flavors I mean. I recommend you Mr. Ahmed that I follow for more detail.

**Reply by amano-kenji** [reply]
Android does application sandboxing. Linux distros don't. The distros I used don't. Fedora doesn't. Gentoo doesn't. Arch doesn't. Ubuntu doesn't.

These distros just let applications do everything.... as effective user IDs. I don't see "any" sandboxing done by these distros.

However you define application sandboxing, they don't "mandate" any sandboxing for all applications.

**Reply by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [reply]
Dear @amano-kenji, I do not want to offend you and you are right default sandboxing does not exist in some flavors=distros, and still hypothetical even if you could setid and setgid to 0, the environment could block most of the access with SELinux, because: adds a Mandatory Access Control (MAC) layer on top of DAC. MAC doesn’t care what the UID is: it enforces policies based on security contexts and rules that are defined by the administrator and checked after DAC. If the policy doesn’t explicitly allow an action, it is denied — even for root. https://docs.redhat.com/de/documentation/red_hat_enterprise_linux/7/html-single/selinux_users_and_administrators_guide/index?utm_source=chatgpt.com . Thank you for your time and bye.

**Reply by topimiettinen** [reply]
@Ghost-in-the-RLAIF-GaN-DNA-code-shell SElinux policies could in theory indeed control access to resources in more fine grained way than what is possible with mount namespaces and capabilities which Firejail uses. Also other MACs aren't so detailed. On the downside, SELinux can't do kernel system call level of access control like seccomp and mounting a tmpfs may be better solution than restricting access. Network support is non-trivial to enable. Tightening the policies locally is difficult because any earlier allow rules can't be deleted by later rules and file system labels are global.

In reality, the SElinux policies as supplied by RedHat are pretty weak: all use apps run in a shared security domain (unrestricted). Upstream Reference Policy (used by Debian) has only a few app policies more (mozilla, chromium, thunderbird etc) and then there's unprivileged user domain (not enabled by Debian). This can be contrasted with Firejail's thousand profiles which are immediately usable.

**Reply by amano-kenji** [reply]
Sandboxing should be mandated by operating system, and users shouldn't have to think about sandboxing at all.

SELinux is merely a bandaid and is very complex. Firejail is at least straightforward for major applications like web browsers.

```
firejail firefox
```

Even firejail is a bandaid for lack of mandatory OS-level application sandboxing that users don't even have to think about.

On android, even grandmas get app sandboxing without having to think about sandboxing. They don't even know what sandboxing means.

When even monkeys get sandboxing without thinking about it, we have "real" sandboxing. When cats, dogs, and chimpanzees play with android phones, they get sandboxing, but they don't know what sandboxing is.

If cats don't get sandboxing, you don't have sandboxing.

You don't scale civilization by forcing people to think about low-level details like sandboxing. Can you scale civilization if you have to constantly think about how water heat exchangers work in the basement of your apartment building? Can you scale civilization when you have to research how to make a sandwich or any food from its raw ingredients instead of just buying it? Making a sandwich from its ingredients takes at least 6 months because you have to grow chickens, fetch water, make bread, and so on. Making a simple pencil takes a million people spread across the globe.

If you are forced to make a pencil from scratch, can you really do things that you do now? You have to mine materials by your hand.

You scale civilization by making things invisible and making things disappear.

If sandboxing really works, we shouldn't even have to know about it until we want to do advanced stuff that cats, monkeys, chimpanzees, and grandmas can't do.

Even chimps get sandboxing on android phones, damn it.

If sandboxing really works, it should be invisible to us unless we want it to be visible.

Don't tell me about MAC, DAC, SELinux, and so on. If monkeys have to know about it, it is broken. Low-level details should disappear.

**Reply by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [reply]
@topimiettinen, dear I thank you for your time and explanation, you are both a kind of convincing, but @amano-kenji  we may be talking past each other slightly.Regarding:"all use apps run in a shared security domain (unrestricted)." It is literally explained in:'https://docs.redhat.com/de/documentation/red_hat_enterprise_linux/7/html-single/selinux_users_and_administrators_guide/index?utm_source=chatgpt.com' , why it doesn't matter that they all do:"run in a shared security domain" ! If I check how many processes are not confined on a Fedora. 
I am pretty certain that it is almost as safe as a sandbox, because despite it’s true that most desktop applications on Fedora run in the unconfined_t domain by default. That means SELinux is not providing per-application sandboxing in the same way Android, seccomp-based sandboxes, or Firejail profiles do. I fully agree with that point. However, that does not mean “there is no sandboxing” or that everything is effectively unrestricted.
SELinux still enforces a system-wide Mandatory Access Control (MAC) layer on top of DAC. 
Even when processes share the unconfined_t domain: The confusion usually stems from the Targeted Policy approach. 
Fedora chooses to protect the system from the apps, whereas Android uses SELinux to protect apps from each other.
To find common ground, we could frame it like this:
- Fedora Desktop: Uses MAC for system-level integrity (protecting the OS core), but lacks per-app isolation by default for usability.
- Android: Uses MAC for full application-level isolation (SEAndroid) by assigning unique domains/categories to every app.
- Firejail/Flatpak: Uses namespaces and seccomp to create user-space containers.

It’s not a binary "sandbox vs. no sandbox" situation; it’s a difference in Threat Models. Fedora protects the house from the guests; Android gives every guest a locked room.
 So the relevant question isn’t “are apps in separate domains?” but rather: What does unconfined_t actually allow?

On Fedora, unconfined_t is broad==true, but it is not equivalent to disabling SELinux entirely. Critical services (e.g., sshd_t, httpd_t, container_t, etc.) are confined. System integrity is still protected by type enforcement on labeled resources.

That is fundamentally different from a pure DAC-only system.
Now, I agree with @top that:
- SELinux does not do syscall filtering like seccomp.
- It is not a mount namespace sandbox.
- The default desktop policy is not tuned for strict per-app isolation.
- Tightening policy locally is difficult.
- But that is a policy choice, not a limitation of SELinux as a mechanism.
Android is in fact built on SELinux, and it achieves per-app isolation by assigning each app its own domain and UID. Fedora could theoretically do the same, it simply chooses not to for usability reasons.

So I would frame it this way:
Desktop Fedora: system integrity MAC, limited per-app isolation.
Android: per-app MAC domains + UID separation.
Firejail: opt-in user-space sandbox using namespaces, seccomp, capabilities.
These are different threat models, not “sandbox vs no sandbox.”

**Reply by amano-kenji** [reply]
In my opinion, there should be a "simple" sandboxing model and an operating system that applies it to every application, and every application "may" programmatically interact with the OS sandbox.

It must be "simple". SELinux is not simple for administrators. Firejail is simpler than SELinux, but firejail is also not simple enough. I don't think people can maintain SELinux if the human civilization takes a hit. Software has become so complex that we need to simplify the entire stack.

Jonathan Blow mentioned creating an operating system from scratch. It is likely not going to be linux, BSD, mac, or any existing operating system.

A completely new operating system design that "comes" with a simple sandboxing model from the "beginning".

Write everything from scratch again with completely new models. Either, don't require any "third-party" application profile, or force application developers to write application profile, or don't require any application profile at all.

You can try to improve sandboxing on existing kernels and existing operating systems, but the return on investment has already reached diminishing returns. It's so complex that users just say they would rather have their privacy violated on mac os if that means they can do productive work faster.

Perhaps, creating an entirely new operating system is impossible in the current human condition, but it may be possible in a post-scarcity economy where many people are free to do whatever they want to do.

**Reply by topimiettinen** [reply]
I like Android's automatic and tight sandboxing model. It works well because the apps are developed to play in the strict sandbox. There's no need to handcraft the sandboxes for each app.  The apps for Linux (nor Mac and Windows) don't expect any sandboxing and there's always a chance that something breaks with the sandboxing when the apps evolve.

The downside with Android (like Mac and Windows) is that it's not possible to tighten its sandboxing with something like Firejail (or firewalls or custom SELinux rules). Hopefully Android for PC will be usable at some point with more tinkering possibilities.

**Reply by topimiettinen** [reply]
> It’s not a binary "sandbox vs. no sandbox" situation; it’s a difference in Threat Models. Fedora protects the house from the guests; Android gives every guest a locked room. So the relevant question isn’t “are apps in separate domains?” but rather: What does unconfined_t actually allow?
> 
> On Fedora, unconfined_t is broad==true, but it is not equivalent to disabling SELinux entirely. Critical services (e.g., sshd_t, httpd_t, container_t, etc.) are confined. System integrity is still protected by type enforcement on labeled resources.

For some system services, yes. But the user apps aren't in separate domains and therefore for example Firefox has full access to all  files in your home directory. The SELinux rules for `unconfined_t` domain don't stop Firefox from writing to your `.bashrc`. So I can't say that the SELinux sandboxing as supplied by Fedora/RH is very strong for user apps. The approach is a compromise aiming for low maintenance effort for RH and the system administrators.

Of course nothing would stop some party from writing tight app-specific SELinux rules for the 1000s of user apps and keeping them up to date when the apps happen to change, like the awesome Firejail team does, but so far this hasn't been seen.

**Reply by Ghost-in-the-RLAIF-GaN-DNA-code-shell** [reply]
@topimiettinen and @amano-kenji ,dears despite I do not recommend flatpak(flathub), because https://flatkill.org/ is a good summary about it, I do use snap and recommend https://documentation.ubuntu.com/snapcraft/stable/. Hope that is a helpful advice and you tell me how secure you consider it,https://snapcraft.io/docs/how-to-guides/manage-snaps/ ?
I.e. you can remove most permissions=connections with i.e. @debian:~$ sudo snap disconnect Application:connection !

**Reply by amano-kenji** [reply]
> Of course nothing would stop some party from writing tight app-specific SELinux rules for the 1000s of user apps and keeping them up to date when the apps happen to change, like the awesome Firejail team does, but so far this hasn't been seen.

I don't think this is scalable. "Third-party" app profiles cannot scale and must go away. Every application should be put in a "simple" fixed sandboxing model.

First-party app profiles could be configured with a manifest file or programmatically, but I think a simple sandboxing model shouldn't require even first-party app profiles.
